compiler/rustc_trait_selection/src/solve/assembly/mod.rs

```diff
@@ -112,6 +110,9 @@ pub(super) trait GoalKind<'tcx>:
     // Consider a clause, which consists of a "assumption" and some "requirements",
     // to satisfy a goal. If the requirements hold, then attempt to satisfy our
     // goal by equating it with the assumption.
+    //
+    // FIXME: change this function to take an actual `Clause` as assumption instead
+    // of a predicate.
```

Comment:
I don't think we should land this without first separately handling that FIXME.

With `NormalizesTo` as a goal getting proven by `Projection` in the environment it's otherwise far too easy to accidentally put a `NormalizesTo` into the environment which is useless. Even without bugs from this, it is still very confusing.